cmd/vm/switch_linux.go

```diff
@@ -0,0 +1,273 @@
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	"github.com/sirupsen/logrus"
+	"github.com/songgao/packets/ethernet"
+	"github.com/songgao/water"
+	"github.com/vishvananda/netlink"
+	"gvisor.dev/gvisor/pkg/tcpip/header"
+)
+
+var (
+	debug    bool
+	tapIface string
+)
+
+const (
+	defaultTapDevice = "eth0"
+	defaultMacAddr   = "5a:94:ef:e4:0c:ee"
+	maxMTU           = 4000
+)
+
+func main() {
+	flag.BoolVar(&debug, "debug", true, "enable debug flag")
+	flag.StringVar(&tapIface, "tap-interface", defaultTapDevice, "tap interface name, eg. eth0, eth1")
+	flag.Parse()
+
+	if debug {
+		logrus.SetLevel(logrus.DebugLevel)
+	}
+
+	// the FD is passed-in as an extra arg from exec.Command
+	// of the parent process. This is for the AF_VSOCK connection that
+	// is handed over from the default namespace to Rancher Desktop's
+	// network namespace, the logic behind this approach is because
+	// AF_VSOCK is affected by network namespaces, therefore we need
+	// to open it before entering a new namespace (via unshare/nsenter)
+	connFile := os.NewFile(uintptr(3), "vsock connection")
+	defer connFile.Close()
+
+	logrus.Debugf("using a AF_VSOCK connection file from default namespace: %v", connFile)
+
+	// this should never happen
+	if err := checkForExsitingIface(tapIface); err != nil {
+		logrus.Fatal(err)
+	}
+
+	tap, err := water.New(water.Config{
+		DeviceType: water.TAP,
+		PlatformSpecificParams: water.PlatformSpecificParams{
+			Name: tapIface,
+		},
+	})
+	if err != nil {
+		logrus.Fatalf("creating tap device %v failed: %s", tapIface, err)
+	}
+	logrus.Debugf("created tap device %s: %v", tapIface, tap)
+	defer func() {
+		tap.Close()
+		logrus.Debugf("closed tap device: %s", tapIface)
+	}()
+
+	if err := linkUp(tapIface, defaultMacAddr); err != nil {
+		logrus.Fatalf("setting mac address [%s] for %s tap device failed: %s", defaultMacAddr, tapIface, err)
+	}
+	if err := loopbackUp(); err != nil {
+		logrus.Fatalf("enabling loop back device failed: %s", err)
+	}
+
+	logrus.Debugf("setup complete for tap interface %s(%s) + loopback", tapIface, defaultMacAddr)
+
+	// catch user issued signals
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, os.Interrupt, syscall.SIGTERM, syscall.SIGINT)
+
+	// try every second until we get DHCP
+	retryTicker := time.NewTicker(time.Second)
+	for {
+		ctx, cancel := context.WithCancel(context.Background())
+		select {
+		case s := <-sigChan:
+			logrus.Errorf("signal caught: %v", s)
+			cancel()
+			os.Exit(1)
+		case <-retryTicker.C:
+			if err := run(ctx, connFile, tap); err != nil {
+				logrus.Error(err)
+			}
+		}
+	}
+}
+
+func run(ctx context.Context, connFile io.ReadWriter, tap *water.Interface) error {
+	errCh := make(chan error, 1)
+	go tx(ctx, connFile, tap, errCh, maxMTU)
+	go rx(ctx, connFile, tap, errCh, maxMTU)
+	go func() {
+		if err := dhcp(ctx, tapIface); err != nil {
+			errCh <- fmt.Errorf("dhcp error: %w", err)
+			ctx.Done()
```

Comment:
I don't understand this line. `ctx.Done()` returns a channel that will emit when the context is done, but you're dropping the return value on the floor. (It _doesn't_ cause the context to be cancelled; for that we need a different `context.WithCancel()`.)

Reply:
I created the `context.WithCancel()` I called the wrong function mistakenly. I will pass in the `cancel` func.